blackbox_repository/tabulated_benchmark.py

```diff
@@ -111,8 +118,25 @@ def blackbox(self) -> Blackbox:
                 self._blackbox = load(self.blackbox_name)
             else:
                 self._blackbox = load(self.blackbox_name)[self.dataset]
+            if self._surrogate is not None:
+                self._blackbox = add_surrogate(self._blackbox, surrogate=self._surrogate)
+
         return self._blackbox
 
+    def __getstate__(self):
+        # we serialize only required metadata information since the blackbox data is contained in the repository and
+        # its raw data does not need to be saved.
+        return {
+            'blackbox_name': self.blackbox_name,
+            'dataset': self.dataset,
+            'surrogate': self._surrogate,
+        }
+
+    def __setstate__(self, state):
```

Comment:
I'd like to learn here. Can you explain to me how Python __setstate__ works? OK, so I want to de-serialize an object, and I have stored 'state'. How do I get 'self' here? Does this call __init__ before calling __setstate__ ?

Reply:
Of-course, when pickle/dill is called to serialize `getstate` is called recursively on the object (conversely `setstate` is called when deserializing). This allows to override the serialization mechanism to store any object for instance with json or dictionary. This is a mechanism that we could use in case we want to save schedulers with a specific format, this is only required if we want to save only required information (e.g. there is no need if all fields are basic types that we all want to store).

> How do I get 'self' here? Does this call init before calling setstate ?

When the object is deserialized, `setstate` is called and it is your responsability to set fields in `self`.

Hope it helps, here is another explanation with minimal examples https://stackoverflow.com/questions/1939058/simple-example-of-use-of-setstate-and-getstate we can chat about this offline if needed.